src/pykeen/nn/message_passing.py

```diff
@@ -515,28 +544,33 @@ def reset_parameters(self):  # noqa: D102
     def forward(
         self,
         x: torch.FloatTensor,
-        source: torch.LongTensor,
-        target: torch.LongTensor,
+        edge_index: Tuple[torch.LongTensor, torch.LongTensor],
         edge_type: torch.LongTensor,
-        edge_weights: Optional[torch.FloatTensor] = None,
     ):
         """
         Calculate enriched entity representations.
 
         :param x: shape: (num_entities, input_dim)
             The input entity representations.
-        :param source: shape: (num_triples,)
-            The indices of the source entity per triple.
-        :param target: shape: (num_triples,)
-            The indices of the target entity per triple.
+        :param edge_index:
+            A tuple of the source and target indices,both with shape (num_triples,)
         :param edge_type: shape: (num_triples,)
             The relation type per triple.
-        :param edge_weights: shape: (num_triples,)
-            Scalar edge weights per triple.
 
         :return: shape: (num_entities, output_dim)
             Enriched entity representations.
         """
+        source, target = edge_index
+
+        if self.edge_weighting is not None and source.numel() > 0:
+            edge_weights = torch.empty_like(source, dtype=torch.float32)
+            for r in range(edge_type.max().item() + 1):
+                mask = edge_type == r
+                if mask.any():
+                    edge_weights[mask] = self.edge_weighting(source[mask], target[mask])
```

Comment:
Schlichtkrull et al. describe this approach in their paper, but mention in the experimental section that they observed a normalization across relations to work better in all settings.